Markdown preview action

kierangilliam opened the pull request:
This PR adds markdown rendering to Zed.

`crates/markdown_preview` exposes a new "markdown: open preview" action. This allows users to preview a rendered version of whichever markdown file they are currently viewing.

https://github.com/zed-industries/zed/assets/18583882/6fd7f009-53f7-4f98-84ea-7dd3f0dd11bf


This PR extends the work done in `crates/rich_text` to render markdown to also support:

- Variable heading sizes
- Markdown tables
- Code blocks
- Block quotes

## Release Notes

- Added `Markdown: Open preview` action to partially close ([#6789](https://github.com/zed-industries/zed/issues/6789)).

## Known issues that will not be included in this PR

- Images.
- Nested block quotes.
- Footnote Reference.
- Inline code highlighting (this will need to be implemented in `rich_text`)
- Checkboxes (`- [ ]` and `- [x]`)
- Syntax highlighting in code blocks.
- Markdown table text alignment.

Original PR: https://github.com/zed-industries/zed/pull/6883

kierangilliam committed:
working in a modal without scroll

kierangilliam committed:
make markdown preview open in a pane

kierangilliam committed:
revert unrelated changes

kierangilliam committed:
clippy

kierangilliam committed:
update preview as new edits come into the buffer

kierangilliam committed:
scrollable preview

kierangilliam committed:
md renderer

kierangilliam committed:
consolidate cmark dependency

kierangilliam committed:
remove unused deps

kierangilliam committed:
add updated lockfile

SomeoneToIgnore commented on crates/markdown_preview/Cargo.toml:
```diff
@@ -0,0 +1,32 @@
+version = "0.1.0"
+edition = "2021"
+publish = false
+license = "GPL-3.0-only"
```
In all other files it's `license = "GPL-3.0-or-later"` now, let's change that

SomeoneToIgnore commented on crates/markdown_preview/src/markdown_preview.rs:
```diff
@@ -0,0 +1,14 @@
+pub mod markdown_preview_view;
+pub mod markdown_renderer;
+
+actions!(markdown, [OpenPreview, SubmitFeedback]);
```
I do not think we need `SubmitFeedback`.

SomeoneToIgnore commented on crates/markdown_preview/src/markdown_preview_view.rs:
```diff
@@ -0,0 +1,169 @@
+
+impl Render for MarkdownPreviewView {
+    fn render(&mut self, cx: &mut ViewContext<Self>) -> impl IntoElement {
+        // TODO: This is wrong but I'm unsure how
```
This is one of the worst place in gpui for me too, here's something in the same `syntax_tree_view` that might help: https://github.com/zed-industries/zed/blob/843919dd1448684b308977e7b3855f853e9afe8f/crates/language_tools/src/syntax_tree_view.rs#L361-L369

There, we wrap an element in something that has to have its bounds calculated, so we have a way to keep it generic and dynamic.

kierangilliam replied:
Made these changes, thank you. 

> we wrap an element in something that has to have its bounds calculated

Am I understanding this right?

1. My `render` function returns a `canvas` with `height=100%` and `width=100%`.
2. Taffy lays out the `canvas` appropriately.
3. When GPUI goes to paint the canvas element, another layout happens for my markdown contents (`paint_callback`).
4. Another paint is triggered (a frame later?)

SomeoneToIgnore replied:
I believe you're right except that both canvas' and element's layouting and paints happen at once, no "a frame later" happens for any paint/delay otherwise we would see it blinking on resize.

SomeoneToIgnore reviewed (commented):
Coming back from https://github.com/zed-industries/zed/discussions/6886#discussioncomment-8276133 with all the answers, but did not browse through the entire PR yet (anybody else is welcome to do so, I plan to do it tomorrow only).

Quick question: this PR looks better than https://github.com/zed-industries/zed/pull/6883 so I wonder if we could close the old one? Or is there anything of value there still?

kierangilliam committed:
update license

kierangilliam committed:
copy layout method from syntax tree view to calculate pane height

kierangilliam committed:
remove TODO

kierangilliam committed:
better md styling + bug fixes

kierangilliam committed:
refactor

kierangilliam committed:
remove unused code

kierangilliam committed:
lints

maxdeviant commented on crates/markdown_preview/LICENSE-GPL:
Could we replace this license file with a symlink to `LICENSE-GPL` in the repository root, like we do with our other licenses?

maxdeviant commented on crates/markdown_preview/src/markdown_preview_view.rs:
```diff
@@ -0,0 +1,137 @@
+            .bg(cx.theme().colors().editor_background)
+            .p_4();
+
+        for item in render_markdown(&self.contents, &self.languages, cx).into_iter() {
```
I think we could replace this loop with a `.children` call, unless I'm missing something.

kierangilliam replied:
I hadn't seen that yet. Handy!

maxdeviant commented:
@kierangilliam This is looking good! I left a few comments, but I think we're getting close to being able to merge this.

kierangilliam committed:
use .children

kierangilliam committed:
remove license

kierangilliam committed:
symlink license

kierangilliam committed:
symlink license

kierangilliam committed:
symlink license

kierangilliam committed:
symlink license

mikayla-maki commented:
This is looking really good! I was poking around to see if we could cache the markdown parsing or keep from serializing the buffer on every keystroke but it seems like both would be a whole project unto themselves. I think this can ship without either getting in the way too much :)

SomeoneToIgnore reviewed (changes requested):
I see a few things that bother me still:

* minor: regular text font size seems to be even bigger in the render, bloating it a bit

* minor: markdown headers in the code have the color, but the rendered version does not

* major: single line code blocks do not seem to rendered at all (but multiline are fine)
<img width="1728" alt="image" src="https://github.com/zed-industries/zed/assets/2690773/27907eba-0ea0-4016-a8f9-1b94f4608b6c">

* major: inner markdown links do not work, erroring (HTTP URLs work just fine)
<img width="1054" alt="image" src="https://github.com/zed-industries/zed/assets/2690773/5945079b-f324-45f7-8fee-366d843b8142">

* major: there's no scrolling happening (presumably, also related to the previous bullet).
If you open the syntax tree view, you can notice both the editor and the view are scrolled when you click in either of the view/editor.
I think we need to have something even more advanced, presumably scrolling always, so both the editor and the render never show different areas.

On the one hand, it's been quite a while we kept this PR unmerged and it got a good progress (for the first try, especially).
On the other hand, it's clearly not finished and needs more work and if we just merge the PR without fixing those, we'll get more tech debt right away.

So, I would try and fix the code blocks highlighting and disable the inner link clicking for now as a minimum before merging this — all other things seem good to have, but maybe later.

mikayla-maki commented:
Did a bit of following up on the inline code highlighting issue, as I'd love to have this feature, and it seems there's a completely separate kind of markdown parsing that Zed also uses, in `crates/language/src/markdown.rs`. This one also seems to do language highlighting, so that Rust code looks correct. I think it would be good to refactor this solution to use that code path, so that we can get great code highlighting in the markdown preview.

kierangilliam commented:
I have more work to do so I'm moving this back to draft. The feedback is very appreciated <3

To address @SomeoneToIgnore comments:

> major: single line code blocks do not seem to rendered at all (but multiline are fine)

Inline code blocks are not supported by `rich_text` which is why they don't render. I addressed this in the PR description but I'll clarify my reasoning here. I'm using `rich_text` to render anything in a paragraph tag. I wanted to make use of existing code where possible to reduce the surface area of this feature. My first PR was simply a wrapper around `rich_text`. I moved more work to `markdown_preview` so that I could support tables and variable heading sizes. 

The more we talk about this feature, the more I'm convinced that I should break away from using `rich_text`!

> major: inner markdown links do not work, erroring (HTTP URLs work just fine)

Yikes. I should've caught that. I'll probably have to move off of using `rich_text` so that I can better control things like this. 

> minor: markdown headers in the code have the color, but the rendered version does not

I don't have an opinion on what the color should be so I'm open to making this change! Is there a public style guide that I can refer to in the future?

> major: there's no scrolling happening... If you open the syntax tree view, you can notice both the editor and the view are scrolled when you click in either of the view/editor.

I don't think I'll have time to address this one in this PR but I'll keep it in mind for a future contribution!

kierangilliam commented:
> it seems there's a completely separate kind of markdown parsing that Zed also uses, in crates/language/src/markdown.rs

@mikayla-maki thanks for pointing this out. This seems to be roughly equivalent to `rich_text`. I'll make myself more familiar with this module. From first glance, it looks like it's purpose is inline documentation. I could extend that to support tables, block quotes, etc, but I'm not sure if that's the right approach. Thoughts / preference?

SomeoneToIgnore commented:
@kierangilliam roger about the highlights, no worries about them now then.
Please do fix the local document link clicking (or at least disallow those in the beginning?) and we're good to merge.

SomeoneToIgnore commented:
Because there's no plugin system to start with, and people seem to want features nonetheless.
The roots of the plugin system that's emerging in https://github.com/zed-industries/zed/issues/7096 are focused on the language integration, and markdown rendering might need more than just the LSP & syntax highlighting support — the editor plugin API which is a big task on its own.

Hence, for now we try to keep the balance between rejecting everything and keeping the good bits in.
A rule of thumb so far is: if there's anything in the Zed repo related to the feature, and the feature looks good per se, we might want to merge it in for now.

kierangilliam committed:
Merge branch 'main' into kieran/md-preview-renderer

kierangilliam committed:
fix cargo deps

kierangilliam committed:
rich_text: only allow external links

kierangilliam commented:
> Please do fix the local document link clicking (or at least disallow those in the beginning?) and we're good to merge.

Done. I'll follow up in another PR with more markdown support while trying to hit some of the other issues you all raised @SomeoneToIgnore.

SomeoneToIgnore reviewed (approved):
Great, thank you for planting the roots!

And thank you for the will to follow-up: I think, we do not fix all those things at once, just start with one and make it good.
So far, seems that tuning/replacing the md renderer seems like the most beneficial one?

burakssen commented:
I am a recent user of zed and I have encountered a problem with markdown preview feature. It basically doesn't render the images. There is only a blank space.

<img width="1624" alt="SCR-20240615-jqkd" src="https://github.com/zed-industries/zed/assets/43742977/08ad3be7-16ba-4675-9008-1343794b7db4">
